### ISSUE TITLE
陈瑾晗任务1提交

### ISSUE THREAD

**Issue by pomelo597** [slwang-ustc/USTC_INT_INFER#1]
[llama.py.txt](https://github.com/user-attachments/files/16942519/llama.py.txt)
[test.py.txt](https://github.com/user-attachments/files/16942521/test.py.txt)

**Comment by Tomorrowdawn** [comment]
将你的代码push到你的仓库中，发branch链接。

**Comment by Tomorrowdawn** [comment]
为什么关闭？

**Comment by pomelo597** [comment]
[https://github.com/pomelo597/USTC_INT_INFER/tree/pomelo597-task1%3Dllamascratch](url)
这样交对吗师兄QAQ

**Comment by pomelo597** [comment]
[https://github.com/pomelo597/USTC_INT_INFER/tree/pomelo597-task1%3Dllamascratch](url)

**Comment by pomelo597** [comment]
[https://github.com/pomelo597/USTC_INT_INFER/tree/pomelo597-task1%3Dllamascratch](url)

**Comment by pomelo597** [comment]
https://github.com/pomelo597/USTC_INT_INFER/tree/pomelo597-task1%3Dllamascratch

**Comment by Tomorrowdawn** [comment]
verified.

注意main分支已经更新很多。记得同步。

可以看一下main里面的git开发流程视频。

按理来说这个issue应该是我来关闭。

**Comment by Tomorrowdawn** [comment]
你的llama py似乎没有实现kv cache

**Comment by pomelo597** [comment]
实现了kv-cache以及任务1和任务2，在branch-jhchen中提交了

**Comment by Tomorrowdawn** [comment]
batch写错了. 对于打包起来的不同句子, attention mask每行必然是不一样的. padding也没有考虑左右pad问题

**Comment by pomelo597** [comment]
师兄我打印输出了一下，attention_mask对于每行是不一样的，temperature采样方法里attention_mask is None是不成立的，因为通过attention_mask = prompt_ids['attention_mask'].to(self.model.device)获取并且传递了。但是确实没有考虑左右pad问题，我再去考虑考虑

**Comment by pomelo597** [comment]
师兄我打印输出了一下，attention_mask对于每行是不一样的，temperature采样方法里attention_mask is None是不成立的，因为通过attention_mask = prompt_ids['attention_mask'].to(self.model.device)获取并且传递了。总之确实是产生了和原始模型一样的结果。但是确实没有考虑左右pad问题，我再去考虑考虑

**Comment by Tomorrowdawn** [comment]
我懂了. 我建议把分词时产生的那个mask叫做padding mask. 尽管transformers这么用了, 但是和通行理解: attention mask是加在attention logits上的mask 不一致. 

另外有一个冷知识: 分词结果可以直接to device而不需要访问内部数据)

没有pad和标准模型一样大概是用标准模型时, tokenizer也忘了pad(

要凸显出这个问题应该需要用两条完全失衡的样本, 譬如说一个长度16, 一个长度2048, 这个时候过多错误的pad token就会影响到输出了.
